airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -34,11 +34,30 @@ class TriggerRuleDep(BaseTIDep):
     IGNOREABLE = True
     IS_TASK_DEP = True
 
+    @staticmethod
+    @provide_session
+    def _get_states_count_upstream_ti(ti, finished_tasks, session):
+        """
+        This function returns the states of the upstream tis for a specific ti in order to determine
+        whether this ti can run in this iteration
+
+        :param ti: the ti that we want to calculate deps for
+        :type ti: airflow.models.TaskInstance
+        :param finished_tasks: all the finished tasks of the dag_run
+        :type finished_tasks: list[airflow.models.TaskInstance]
+        """
+        if not finished_tasks:
+            # this is for the strange feature of running tasks without dag_run
+            finished_tasks = [t for t in ti.task.dag.get_task_instances(start_date=ti.execution_date,
+                                                                        end_date=ti.execution_date)
```

Comment:
We could do the filtering in the DB here?

```suggestion
                                                                        end_date=ti.execution_date,
                                                                        state=(State.finished() + [State.UPSTREAM_FAILED])
```

Reply:
And we should pass the same `session` down to `ti.task.dag.get_task_instances)`